### DISCUSSION TITLE
deceleration rate

### DISCUSSION THREAD

**Discussion by opndesk** [joshua-8/Derivs_Limiter#12] [Q&A] [answered]
do you have plans to include a separate rate for accel and decel ?

**Comment by joshua-8** [comment, marked as answer]
Thank you for your interest in this library, and for your suggestion. I hadn't thought of having different rates for acceleration and deceleration, but I can see how that could be useful. It didn't actually take much time to separate accel and decel, A version of the library with separate accel and decel can be downloaded [here](https://github.com/joshua-8/Derivs_Limiter/tree/884ee9500a4b3a9e9dedf30f88f3f8a205057176). 

Derivs_Limiter.setAccelLimitActuallyAcceleration() sets accel while every other reference to acceleration limits is actually deceleration.

I don't want to make a new release yet since I want to think about the proper way to add this option instead of this quick hack.

Here is an example showing velocity increasing at a slower rate than it decreases at. The updated [Derivs_Limiter.h](https://raw.githubusercontent.com/joshua-8/Derivs_Limiter/884ee9500a4b3a9e9dedf30f88f3f8a205057176/src/Derivs_Limiter.h) can be copied into a second tab in the Arduino IDE or pasted into the top of the editor (or the Derivs_Limiter folder can be put in Documents/Arduino/libraries. (let me know if you have issues getting it to work)

```
#include "Derivs_Limiter.h"

Derivs_Limiter limiter = Derivs_Limiter(60, 100); // velocityLimit, decelerationLimit
void setup()
{
  limiter.setAccelLimitActuallyAcceleration(35); //this sets the accel limit to something different from the "accel limit" set in other places that is now actually the decel limit.
  Serial.begin(9600);
}
void loop()
{
  if (millis() % 10000 < 5000) { //toggles every 5 seconds
    limiter.calc(0);
  } else {
    limiter.calc(180);
  }
  Serial.print(limiter.getPosition());
  Serial.print(",");
  Serial.print(limiter.getVelocity());
  Serial.println();
}

```

![image](https://user-images.githubusercontent.com/59814881/167264304-83fa42db-cca0-4abc-8539-67f8b62831f1.png)

Please note that these changes have not been tested very thoroughly. Please test it yourself before trusting it, and please let me know if this is what you were looking for or if there's another way that I could help.

**Reply by opndesk** [reply]
yes! it works great for me

**Reply by joshua-8** [reply]
Version 3.0 of this library now supports a different deceleration rate. Here's [an example](https://github.com/joshua-8/Derivs_Limiter/blob/master/examples/SmoothServo/SmoothServo.ino) showing how a Derivs_Limiter object can be made with deceleration as the third parameter.
`Derivs_Limiter(velocityLimit,accelerationLimit,decelerationLimit)`
When I tested the old solution more thoroughly, I found that it could sometimes jitter around and really not work so well, so I would recommend changing to the new version.
If you have time to test version 3, I would appreciate any feedback. Thanks for your interest and your suggestion!
-Joshua

**Reply by opndesk** [reply]
just did play a bit with the samples and I can see the adjustments, will test my application this weekend and send feedback

**Reply by opndesk** [reply]
it works great! what about change of velocity on the fly for next challenge? I call this a complex move profile.
I appreciate you really understand problem solving.
![Screenshot 2022-06-28 104629](https://user-images.githubusercontent.com/33498550/176209707-172007fb-7b86-417e-a070-8cc7d48a7fc3.png)

**Reply by joshua-8** [reply]
Thank you for your kind feedback and doing what looks like a very good test. I would be interested in hearing more about your "velocity on the fly" idea. Something I added to the new version was a velocity control mode where the velocity can either be set to a value immediately and stays at that value, or a velocity target can be set and the velocity smoothly accelerates towards that: 
setVelConstant(): https://joshua-8.github.io/Derivs_Limiter/class_derivs___limiter.html#a55560bcd022d938be042ec4161ce6c4f
setVelTarget(): https://joshua-8.github.io/Derivs_Limiter/class_derivs___limiter.html#ad775094ddc9877e618588f0bb86f5e56

**Reply by opndesk** [reply]
YES! that is exactly what I was talking about. it is amazing very professional. thanks.

**Reply by joshua-8** [reply]
I'm glad it's what you're looking for! Thanks for the encouragement! If you ever have more feedback or questions don't hesitate to ask. Best of luck with all your projects!
